### PR TITLE
Copy of the inputs for the evaluation during compilation

### PR DESCRIPTION
**Pull request by vkovinicTT**
### Summary

This PR ensures that a copy of the input tensors is created when running forward during compilation (e.g. when trying to verify outputs using forward pass of the framework model). This prevents unintended modifications to the original input tensors if in-place operations are performed during the forward pass.

In order for it to work, first this [PR](https://github.com/tenstorrent/tt-tvm/pull/60) in the `tt-tvm` needs to be merged.

### Why is this needed?
Some models perform in-place operations on input tensors during the forward pass, which can lead to unintended changes in the original inputs. By making a copy of the inputs, we ensure correctness and avoid potential issues when running forward during compilation.

Example test:
```python
    class Inplace(torch.nn.Module):
        def __init__(self):
            super().__init__()

        def forward(self, x):
            y = x + 1 
            x += 2 # in-place operation on the input that causes the inputs to change during forge.compile(...)

            return x + y
        
    input = torch.zeros(shape, requires_grad=False)
    framework_input = input.detach().clone()
    tt_inputs = [input]

    framework_model = Inplace()
    y = framework_model(framework_input)
        
    compiled_model = forge.compile(framework_model, sample_inputs=tt_inputs, module_name="inplace")
    tty = compiled_model(*tt_inputs)[0]

    compare_with_golden(golden=y, calculated=tty) # this would fail
```
❗❗❗ IMPORTANT NOTES ❗❗❗

In the training mode if the inputs require grad, pytorch will try to do the forward and will throw runtime error:
`a leaf Variable that requires grad is being used in an in-place operation.`, while the compiled model silently computes the gradients and **this need to be addressed**.

Introducing this change our compiler will not change it's input, so it will act as tensorflow (it doesn't change it's input tensor), but it is won't be aligned with pytorch which allows in-place changes of the input tensor.

### CODE REVIEW

File: forge/test/mlir/operators/tm/test_in_place.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+            super().__init__()
+
+        def forward(self, x):
+            y = x + 1
+            x += 2
+
+            return x + y
+
+    input = torch.zeros(shape, requires_grad=False)
+    framework_input = input.detach().clone()
```
Comment (nvukobratTT) [inline forge/test/mlir/operators/tm/test_in_place.py:31]: Why we're doing detach and clone here?

File: forge/test/mlir/operators/tm/test_in_place.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+    tt_inputs = [input]
+
+    framework_model = Inplace()
+    y = framework_model(framework_input)
+
+    compiled_model = forge.compile(framework_model, sample_inputs=tt_inputs, module_name="inplace")
+    tty = compiled_model(*tt_inputs)[0]
+
+    compare_with_golden(golden=y, calculated=tty)
```
Comment (nvukobratTT) [inline forge/test/mlir/operators/tm/test_in_place.py:40]: Should we use the standard verify function?

File: forge/test/mlir/operators/tm/test_in_place.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+    print(framework_input)
```
Comment (nvukobratTT) [inline forge/test/mlir/operators/tm/test_in_place.py:41]: Do we need these prints?

File: forge/test/mlir/operators/tm/test_in_place.py [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+    framework_model = Inplace()
+    y = framework_model(framework_input)
+
+    compiled_model = forge.compile(framework_model, sample_inputs=tt_inputs, module_name="inplace")
+    tty = compiled_model(*tt_inputs)[0]
+
+    # convert tensor from tf to torch
+    y = torch.tensor(y.numpy())
+
+    compare_with_golden(golden=y, calculated=tty)
```
Comment (nvukobratTT) [inline forge/test/mlir/operators/tm/test_in_place.py:75]: Similar comments as for PT example

File: forge/test/mlir/operators/tm/test_in_place.py (hunk 1 above)
Comment (vkovinicTT) [reply forge/test/mlir/operators/tm/test_in_place.py:31]: Because if run inference on the `torch` module and then want to run it on `compiled module`, inputs for the compiled module would be different due to the changes made by torch (as default torch is not functional, meaning it will change input)

File: forge/test/mlir/operators/tm/test_in_place.py (hunk 2 above)
Comment (vkovinicTT) [reply forge/test/mlir/operators/tm/test_in_place.py:40]: For the nature of the in-place problem we are facing, I have separate input for `compiled module` and `torch module` so I can't pass it to the `verify` function.

File: forge/test/mlir/operators/tm/test_in_place.py (hunk 3 above)
Comment (vkovinicTT) [reply forge/test/mlir/operators/tm/test_in_place.py:41]: Not really.

**Comment by vkovinicTT** [comment]
I would still wait with this PR until we resolve which policy we want to follow (functional or per framework policy).
